### PR TITLE
chore: split CachingFS from trace recorder

### PR DESCRIPTION
**Pull request by dgozman**
So far, no semantic changes at all.

### CODE REVIEW

File: packages/playwright-core/src/server/trace/recorder/tracing.ts [hunk 1]
```diff
@@ -231,32 +228,32 @@ export class Tracing extends SdkObject implements InstrumentationListener, Snaps
...
+    await this._fs.flush();
   }
 
   async stopChunk(params: TracingTracingStopChunkParams): Promise<{ artifact?: Artifact, entries?: NameValue[] }> {
     if (this._isStopping)
       throw new Error(`Tracing is already stopping`);
     this._isStopping = true;
 
-    if (!this._state || !this._state.recording) {
+    const state = this._state!;
```
Comment (yury-s) [inline packages/playwright-core/src/server/trace/recorder/tracing.ts:239]:
```suggestion
    const state = this._state;
```

File: packages/playwright-core/src/server/trace/recorder/tracing.ts [hunk 2]
```diff
@@ -194,7 +191,7 @@ export class Tracing extends SdkObject implements InstrumentationListener, Snaps
   }
 
   private async _changeTraceName(state: RecordingState, name: string) {
-    await this._appendTraceOperation(async () => {
+    await this._fs.flushLockAndRun(async () => {
```
Comment (yury-s) [inline packages/playwright-core/src/server/trace/recorder/tracing.ts:194]: flushLockAndRun => flushAndRun ?

**Review by yury-s — APPROVED** [review]

**Comment by pavelfeldman** [comment]
Let's hold it off for now to prevent potential merging conflicts

File: packages/playwright-core/src/server/trace/recorder/tracing.ts [hunk 3]
```diff
@@ -512,3 +486,70 @@ function createAfterActionTraceEvent(metadata: CallMetadata): trace.AfterActionT
...
+    if (!this._dirsCreated.has(dir)) {
+      await fs.promises.mkdir(dir, { recursive: true }).catch(() => this._hadErrors = true);
+      this._dirsCreated.add(dir);
+    }
+  }
+
+  async appendText(filePath: string, text: string) {
+    return this._appendOperation(async () => {
+      await this._makeDirIfNeeded(filePath);
+      await fs.promises.appendFile(filePath, text).catch(() => this._hadErrors = true);
```
Comment (pavelfeldman) [inline packages/playwright-core/src/server/trace/recorder/tracing.ts:522]: I'm not sure I understand where the caching is coming from.

File: packages/playwright-core/src/server/trace/recorder/tracing.ts [hunk 4]
```diff
@@ -231,44 +226,53 @@ export class Tracing extends SdkObject implements InstrumentationListener, Snaps
...
+    const suffix = state.chunkOrdinal ? `-${state.chunkOrdinal}` : ``;
+    const networkFile = path.join(state.tracesDir, state.traceName + `${suffix}.network`);
+    this._fs.copyFile(state.networkFile, networkFile);
+
     // Chain the export operation against write operations,
-    // so that neither trace files nor sha1s change during the export.
-    return await this._appendTraceOperation(async () => {
-      if (params.mode === 'discard')
+    // so that files that are not modified while being zipped.
+    return await this._fs.flushLockAndRun(async () => {
```
Comment (pavelfeldman) [inline packages/playwright-core/src/server/trace/recorder/tracing.ts:264]: What is Lock?

**Review by pavelfeldman — APPROVED** [review]

**Pull request closed without merging**